### PR TITLE
Tagging of apparent CNLoH in the matched normal

### PR DESCRIPTION
**Pull request by LeeTL1220**
- Adds an option to look for apparent mosaicism CNLoH in the matched normal.  Also, added some parameters for tuning the behavior.  This is still based on hard thresholds and should probably be better evaluated.  Also, the CNLoH in the matched normal is likely an artifact.
- Rename the unsupported `combine_tracks.wdl` to `model_segments_postproessing.wdl`

### CODE REVIEW

**Comment by samuelklee** [comment]
Actually, before I review, can you show me some examples (perhaps over Slack)?  The ModelSegments plots would be fine, if the events are visible there.

**Comment by LeeTL1220** [comment]
@samuelklee Sent via slack

**Comment by samuelklee** [comment]
As we discussed, it's possible that these are simply common germline CNVs that are being median-normalized out in CR by the PoN.  Let's investigate the sample-median-normalized counts in some of the questionable regions, along with the per-bin medians in the PoN.  I do not think a gCNV run is necessary (it will probably be a bit expensive, anyway).

More generally, I think a better approach to germline tagging would be to avoid the caller entirely.  Let's take the ModelSegments output for a normal, and then tag ModelSegments segments in the tumor that sufficiently overlap any normal segment in CR-AF-genomic space (where we have some freedom to define the overlap criteria).  Essentially, let's just try to highlight differences between the tumor and normal in CR-AF space.  This would rescue events in the normal that may be further amplified or deleted in the tumor.  Subsequently, simple filtering of these events would be less misleading than imputation.

I do not think such tagging should be implemented in Java, if we can avoid it.  Rather, a relatively simple python script that runs through each tumor segment and checks for overlaps would suffice.  This script could output a tagged/filtered ModelSegments result, as well as do the conversion step for downstream tools.  This also obviates the need for the Java code for combining segment breakpoints and additional CNV collection classes in the current post-processing tools.  What do you think?

**Comment by samuelklee** [comment]
If these events were indeed not CNLOH, as we discussed, then I don't think we should merge this.  Perhaps we should take a step back and answer definitively whether simply blacklisting common germline regions is enough to most postprocessing.  Should be straightforward to run an evaluation with and without blacklisting---and hopefully our truth data accurately reflects whether blacklisting is desirable.

If tagging/filtering rare germline is still a concern, then I'd say the next step is to see whether simply changing segmentation parameters to artificially decrease resolution and/or simple length-based filtering suffices.

Finally, simple filtering based on CR-AF as described above could be implemented.  If the normal is available, we can make IS_NORMAL calls simply based on the overlap of the ModelSegments posteriors (with corresponding qualities).  If not, then some heuristic determination of the normal state from the tumor alone as in Marton's caller could be performed.  This would combine the IS_NORMAL calling and filtering steps into one simple tool.  The output could be a tagged/filtered ModelSegments .seg file and the corresponding VCF.

**Comment by LeeTL1220** [comment]
@samuelklee 

The CNLoH are probably artifacts (faux-CNLoH) -- I think I misdiagnosed.  However, there are other changes in this PR and we can always disable the faux-CNLoH pruning by default (there is a flag). Additionally, artifact or not, it is useful to remove the faux-CNLoH.   Users are already using this branch and giving me positive feedback (definitely more positive than adjusting num_changepoints_penalty_factor).  I suggest merging mostly for practical reasons.  It buys us more time to put in a principled solution.  And this workflow is clearly marked as an unsupported prototype anyway (as are the GATK CLIs).  I want to emphasize that this whole workflow is not a long-term solution.  In other words, I would like to get this in and then focus on a supported solution.

Two comments:  
> If these events were indeed not CNLOH, as we discussed, then I don't think we should merge this. Perhaps we should take a step back and answer definitively whether simply blacklisting common germline regions is enough to replicate/obviate most of the postprocessing. Should be straightforward to run an evaluation with and without blacklisting---and hopefully our truth data accurately reflects whether blacklisting is desirable.

There are definitely events that get missed without the germline tagging, so this is an improvement over blacklisting alone.  And while I have seen erroneous germline tagging (i.e. false calling a segment germline), it was only ever due to really noisy data (e.g. a bad PoN) or a poorly tuned segment caller.

I am pretty sure that most common germline regions are being blacklisted already.  The hotspots addressed in this PR (faux-CNLoH) could be added, but I think we will find new areas and a few of these areas were rather big.   I have users that are actively using this from the branch, for reasons other than the faux-CNLoH pruning.  Results are improving without an appreciable hit to sensitivity, which we got when using parameters like num_changepoints_penalty_factor.  As a compromise, I can always default the CNLoH piece to `false`, since there are other useful changes on this branch.  (Users did not have as strong an opinion about the faux-CNLoH pruning, since GISTIC does not use MAF and ABSOLUTE requires a manual review).

> simple filtering based on CR-AF as described above could be implemented. If the normal is available, we can make IS_NORMAL calls simply based on the overlap of the ModelSegments posteriors (with corresponding qualities). If not, then some heuristic determination of the normal state from the tumor alone as in Marton's caller could be performed. This would combine the IS_NORMAL calling and filtering steps into one simple tool. The output could be a tagged/filtered ModelSegments .seg file and the corresponding VCF
And this would be a "better solution"  Shall I file an issue for this?  This could also allow us to obviate the TagGermline tool, which is fine by me.

**Comment by samuelklee** [comment]
Executive summary: 

My main concern is that the amount of unsupported code is continuing to grow.  Adding this PR would bring the total to about ~5k lines of WDL, Java, and test code.  In comparison, the amount of corresponding supported CNV code clocks in around ~33k---this includes all of gCNV, as well!  Development time has also been non-negligible and dates back to pre-4.0 release.

Another concern is that the number of users of this unsupported code is also growing.  In fact, it seems like we are actively pointing users to it.  This seems unsustainable going forward.

Finally, I don't think we have satisfactorily demonstrated which of the functions accomplished by this code (format conversion, blacklisting, germline/"CNLOH" tagging) are necessary or cannot be performed by existing code or more streamlined and principled methods.  (Some of these functions, such as IGV conversion, are already performed by existing code.)  Of those functions, I think format conversion is the only one we should retain from this code.  So if this PR introduces a useful GISTIC conversion, no harm in merging it.

This all sounds like a decision for the new tech lead!  @mwalker174 any thoughts?  

More detailed responses follow:

> Users are already using this branch and giving me positive feedback (definitely more positive than adjusting num_changepoints_penalty_factor). I suggest merging mostly for practical reasons. It buys us more time to put in a principled solution. And this workflow is clearly marked as an unsupported prototype anyway (as are the GATK CLIs). I want to emphasize that this whole workflow is not a long-term solution. In other words, I would like to get this in and then focus on a supported solution.

While it's great that users are giving positive feedback, I refer you to CellBender team's manifesto at https://github.com/broadinstitute/cellbender.  Can these users help us definitively resolve whether these events are 1) germline with incorrectly normalized CR, or 2) mosaic CNLOH?  If not, then we have not even taken the first step to correctly identify the issue.  So it seems a bit premature to even prototype a method, much less merge it.

I think this PR, as is, muddies the waters quite a bit.  For example, it introduces a new Record class that denotes this type of "CNLOH" with a `C`.  If we want to merge this, I suggest that we first correctly identify the issue.  If these events are not mosaic CNLOH, then we should clean up all mention of CNLOH in this code.

Either way, can we quantify the level of improvement gained by filtering such events in a reproducible evaluation?  If so, let's bring that into gatk-evaluation.

Finally, there are many more options available to change the segmentation and/or resolution than the single one you mentioned.  If the users you are working with can clearly specify their analysis goals in terms of resolution, then it might be possible to sidestep the problem entirely without adding more unsupported code.  This would also buy us more time to put in a principled solution, without the risk of unsupported code getting entrenched in their workflows.

> There are definitely events that get missed without the germline tagging, so this is an improvement over blacklisting alone. And while I have seen erroneous germline tagging (i.e. false calling a segment germline), it was only ever due to really noisy data (e.g. a bad PoN) or a poorly tuned segment caller.

This is encouraging.  This means that a straightforward approach to germline filtering, such as simply identifying overlapping posteriors as mentioned above, should work well.  Prototyping this approach shouldn't take long at all, especially when the matched normal is guaranteed to be available, as it is in this workflow (tumor-only would require some work to identify the normal state, as mentioned previously).  I'd rather just roll that, evaluate it, and merge it instead.

Key here is that we sidestep the deficiencies of the current CR-only caller, which also shares the blame for this "CNLOH" issue (since these events aren't called in the normal and don't become candidates for tagging, as currently implemented).

> And this would be a possible "better solution" Shall I file an issue for this? This could also allow us to obviate the TagGermline tool, which is fine by me.

I've already expanded the scope of https://github.com/broadinstitute/gatk/issues/4115 to include this.

File: scripts/unsupported/model_segments_postprocessing_cnv/model_segments_postprocessing.wdl [hunk 1]
```diff
@@ -51,45 +53,98 @@
 #  - centromere_tracks_seg:  gs://gatk-best-practices/somatic-hg38/final_centromere_hg38.seg
 #  - gistic_blacklist_tracks_seg:  gs://gatk-best-practices/somatic-hg38/CNV.hg38liftover.bypos.v1.CR1_event_added.mod.seg
 #
-#  Dev note:  FYI...This workflow is easily tested on a laptop.
+#  Dev notes:
```
Comment (mwalker174) [inline scripts/unsupported/model_segments_postprocessing_cnv/model_segments_postprocessing.wdl]: Add an explanation of why CNLoH events and common germline events are not distinguishable with this method.

File: scripts/unsupported/model_segments_postprocessing_cnv/model_segments_postprocessing.wdl [hunk 2]
```diff
@@ -51,45 +53,98 @@
...
     Int? germline_tagging_padding
     String group_id
     String gatk_docker
     Int? max_merge_distance
     Array[String]? annotations_on_which_to_merge
     Int? min_hets_acs_results
     Float? maf90_threshold
-
+    Float maf90_threshold_final = select_first([maf90_threshold, 0.47])
+    Boolean? is_ignore_cnloh_in_matched_normal
```
Comment (mwalker174) [inline scripts/unsupported/model_segments_postprocessing_cnv/model_segments_postprocessing.wdl]: Maybe `is_ignore_possible_cnloh_in_matched_normal` ?

**Comment by droazen** [comment]
@samuelklee @mwalker174 Do either of you want to take over this old PR, or would you prefer to close it?

**Comment by samuelklee** [comment]
Closing for reasons outlined above. We can always reopen or grab the relevant conversion tidbits if needed.

**Pull request closed without merging**